storm-server/src/main/java/org/apache/storm/daemon/nimbus/Nimbus.java

```diff
@@ -4716,4 +4744,206 @@ public IScheduler getForcedScheduler() {
+
+        //Topology metrics distribution
+        private final Histogram topologiesNumTasks = registerHistogram.apply("topologies:num-tasks");
+        private final Histogram topologiesNumExecutors = registerHistogram.apply("topologies:num-executors");
+        private final Histogram topologiesNumWorker = registerHistogram.apply("topologies:num-workers");
+        private final Histogram topologiesUptime = registerHistogram.apply("topologies:uptime-secs");
+        private final Histogram topologiesReplicationCount = registerHistogram.apply("topologies:replication-count");
+        private final Histogram topologiesRequestedMemOnHeap = registerHistogram.apply("topologies:requested-mem-on-heap");
+        private final Histogram topologiesRequestedMemOffHeap = registerHistogram.apply("topologies:requested-mem-off-heap");
+        private final Histogram topologiesRequestedCpu = registerHistogram.apply("topologies:requested-cpu");
+        private final Histogram topologiesAssignedMemOnHeap = registerHistogram.apply("topologies:assigned-mem-on-heap");
+        private final Histogram topologiesAssignedMemOffHeap = registerHistogram.apply("topologies:assigned-mem-off-heap");
+        private final Histogram topologiesAssignedCpu = registerHistogram.apply("topologies:assigned-cpu");
+
+        /**
+         * Constructor to put all items in ClusterSummary in MetricSet as a metric.
+         * All metrics are derived from a cached ClusterSummary object,
+         * expired {@link ClusterSummaryMetricSet#CACHING_WINDOW} seconds after first query in a while from reporters.
+         * In case of {@link com.codahale.metrics.ScheduledReporter}, CACHING_WINDOW should be set shorter than
+         * reporting interval to avoid outdated reporting.
+         */
+        ClusterSummaryMetricSet() {
+            //Break the code if out of sync to thrift protocol
+            assert ClusterSummary._Fields.values().length == 3
+                && ClusterSummary._Fields.findByName("supervisors") == ClusterSummary._Fields.SUPERVISORS
+                && ClusterSummary._Fields.findByName("topologies") == ClusterSummary._Fields.TOPOLOGIES
+                && ClusterSummary._Fields.findByName("nimbuses") == ClusterSummary._Fields.NIMBUSES;
+
+            final CachedGauge<ClusterSummary> cachedSummary = new CachedGauge<ClusterSummary>(CACHING_WINDOW, TimeUnit.SECONDS) {
+                @Override
+                protected ClusterSummary loadValue() {
+                    try {
+                        if (active) {
+                            ClusterSummary newSummary = getClusterInfoImpl();
+                            LOG.debug("the new summary is {}", newSummary);
+                            //Force update histogram upon each cache refresh
+                            //This behavior relies on the fact that most common implementation of Reporter
```

Comment:
1. I think it'd be better to have all metrics in sync. Having a StormTimer is also problematic because there's no way to guarantee when histogram gets updated vs. when reporters decide to report metrics. We would then get out of sync metrics for gauge and histogram, which can be confusing.

2. Since for histogram we track distribution for only one ClusterSummary measurement. The out of sync behavior give us no guarantee if newer value will override older ones that haven't been reported, and it'd be hard to figure out the correct caching window.

This ultimately boils down to the difference between active update pattern of histogram vs. passive update pattern of gauge, and this is the best kind of hack I can think of as of now to mediate the conflict. An alternative would be to convert all histogram to gauge, which means manually calculate avg/min/max/percentile. This will bloat the code greatly and I don't see a very good reason doing it, almost like trading one hack for another.

Reply:
We can also develop a passive histogram that simply wraps a round a gauge. It's better to file another Jira then.

Reply:
I have another potentially related question: What does this gauge output to the reporter? Is it going to be the toString on the ClusterSummary, or what happens when the value is delivered to the reporter? Is the gauge just here to update the other metrics?

Reply:
What gauged are you talking about? Basically my envisioning is that we subclass a histogram that simply wraps around many derivative gauges which computes avg/min/max on their own. They are encapsulated as a ReservoirSnapshot upon the histogram is queried. On the other hand we disable the `update` method so they can't be actively updated.

Reply:
The `cachedSummary` gauge

Reply:
Nevermind, I should read a little closer.

Reply:
This gauge is never registered. It simply serves as a cache for other gauge to use. Otherwise we'd have to come up with our own caching mechanism.

Reply:
How about making a repeating timer task that fetches the cluster summary, updates the histograms, and updates fields that are then read by the gauges? 

The reason I don't like this is that it makes a lot of assumptions about the internals of code we don't control. If the reporter looks at histograms before gauges, you get the issue you mention (and keep in mind anyone can implement a reporter). If the DerivativeGauge were implemented differently (e.g. if the parent gauge pushed the value to the child, rather than the other way around) this also would break.